src/utils/getter/timeseriesData.ts

```diff
@@ -16,7 +16,7 @@ export class TimeseriesRowGetter<T extends TimeseriesRow = TimeseriesRow> extend
 
     get value(): number {
         switch (this.mode) {
-            case "ab":
+            case "ab": {
```

Comment:
https://github.com/FeiBam/wt-data-project.web/pull/1

Reply:
![image](https://user-images.githubusercontent.com/12800094/208992670-d9e77c0b-4a6c-4131-b97c-d0766f957fe5.png)

Is it type safe?

Reply:
no

Reply:
FYI:
https://bobbyhadz.com/blog/typescript-access-object-property-dynamically
https://www.typescriptlang.org/docs/handbook/advanced-types.html#index-types

Reply:
Thanks. That's a great practice for advanced types. The return type is `number` or `string` not `TimeseriesRow`. I modified it, and now it looks not bad.

But it still has a lot of `as`.

```typescript
type TimeseriesRow<T extends Mode> = T extends 'ab'
    ? TimeseriesABRow
    : T extends 'rb'
        ? TimeseriesRBRow
        : T extends 'sb'
            ? TimeseriesSBRow
            : never;

export class TimeseriesRowGetter<M extends Mode, T extends TimeseriesRow<M>, K extends keyof T> extends Getter {
    protected readonly data: T;
    protected readonly mode: M;
    private readonly measurement: Measurement;

    constructor(data: T, mode: M, measurement: Measurement) {
        super();
        this.data = data;
        this.mode = mode;
        this.measurement = measurement;
    }

    get value(): number {
        return this.data[`${this.mode}_${this.measurement}` as K] as number;
    }

    get br(): string {
        return this.data[`${this.mode}_br` as K] as string;
    }

    get lowerBr(): number {
        return +this.data[`${this.mode}_lower_br` as K] as number;
    }
}

```

Reply:
what's the ```typeof this.data[`${this.mode}_${this.measurement}` as keyof T]```?
can TS infer it correctly from `T extends TimeseriesRow<M>`?

Reply:
also u should use [`Number()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/Number) and [`String()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/String) to convert their types in runtime, similar with the previous commit: https://github.com/FeiBam/wt-data-project.web/pull/1/commits/de12004fdf790b33355c4d18eba43a5b542e9a51

Reply:
No. `T` is `TimeseriesRow<M>` so it's not necessary I guess. Still cannot pass the type check.

Reply:
> also u should use [`Number()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/Number) and [`String()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/String) to convert their types in runtime, similar with the previous commit: [FeiBam@de12004](https://github.com/FeiBam/wt-data-project.web/commit/de12004fdf790b33355c4d18eba43a5b542e9a51)

No it's not convertion. It's only for narrowing the type `number | string` to `string` or `number`.

Reply:
I mean that [unary plus `+` operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Unary_plus), it's being commonly used as converting anything to number.
plz choose ur hero: https://stackoverflow.com/questions/17106681/parseint-vs-unary-plus-when-to-use-which
![](https://i.stack.imgur.com/LLrgj.png)

Reply:
Good. Merged.  0f90a2bdc8feed5485b9004758580f9d95d4d35f

CC @kokoro-aya

Reply:
I don't think it's a good idea to use implicit conversion, even if it may work.

Reply:
> I don't think it's a good idea to use implicit conversion, even if it may work.

All columns in this table are explicit conversions, it's just comparing across all these conversion methods.
We're not talking about implicit conversion, for that there's another table: https://thomas-yang.me/projects/oh-my-dear-js/

Reply:
I see, then it may be kind of flavour preference.

Reply:
This table figured out that `Number(x)` is basically the same with `+x`, I would prefer `Number(x)` since it's more clarified.

Reply:
Agreed.

Reply:
and the js code minifier in your building chain should be able to convert `Number(x)` to `+x` for saving bytes over internet, unless u're not using minifier, u don't have to enforce urself to use `+x` for that reason.